### PR TITLE
feat: support LIST_APPEND and LIST_CONCAT

### LINKED ISSUE #897: [Feature] Support common LIST operations
## Motivation

neuG currently supports only a limited set of LIST operations. The built-in function collection registers list creation, element extraction, and containment, while common operations such as appending an element or explicitly concatenating lists are not consistently available as LIST functions.

### `list_append(LIST<T>, T) -> LIST<T>`

Append one element to the end of a list.

```cypher
RETURN list_append([1, 2], 3)
-- [1, 2, 3]
```

### `list_concat(LIST<T>, LIST<T>) -> LIST<T>`

Concatenate two lists. This should provide a function equivalent to the existing `LIST + LIST` behavior.

```cypher
RETURN list_concat([1, 2], [3, 4])
-- [1, 2, 3, 4]
```

Other commonly used LIST operations, such as `list_prepend`, can be added as part of the same LIST-function framework or in follow-up work.

## Expected behavior

- Infer a compatible common child type and insert casts when appropriate.
- Support primitive and nested types, including STRUCT elements.
- Return a clear binder error for incompatible child/element types.
- Define consistent behavior for NULL lists and NULL elements.
- Reuse the existing list-vector execution infrastructure where possible.

## Acceptance criteria

- Register `list_append` and `list_concat` in the built-in function collection.
- Add binder and execution tests for primitive, nested/STRUCT, empty, and NULL inputs.
- Test compatible implicit casts and incompatible types.
- Validate the existing large-list-of-STRUCT `list_append` e2e case in `tests/e2e/queries/projection/single_label.test`.
- Keep `list_concat(a, b)` behavior consistent with `a + b`.

### PR DESCRIPTION
**Pull request by shirly121**
## Summary

- add built-in `list_append` and `list_concat` functions for LIST and ARRAY inputs
- infer a common element type and support compatible casts, empty lists, and NULL values
- preserve existing list concatenation behavior while returning variable-length LIST results
- add Cypher manual documentation and Python end-to-end coverage for literals, properties, parameters, errors, and NULL storage behavior

## Testing

- `make format-check`
- `python3.9 -m pytest -q tests/test_list_func_docs.py tests/test_db_list.py -k 'list_append_and_concat or documentation_example'`

Closes #897

### CODE REVIEW

File: doc/source/cypher_manual/expression/list_func.md [hunk 1]
```diff
@@ -0,0 +1,339 @@
+# List Functions
+
+Since v0.2.0, NeuG provides built-in functions for working with list-like values. These functions provide common operations for constructing, combining, and manipulating collection values in Cypher queries.
+
+The currently supported list functions are summarized below.
+
+| Function                          | Description                            | Example                              |
+| --------------------------------- | -------------------------------------- | ------------------------------------ |
+| `list_append(list_like, element)` | Appends one element to a list or array | `RETURN list_append([1, 2], 3)`      |
```
Comment (longbinlai) [inline doc/source/cypher_manual/expression/list_func.md:9]: 可以是 list of list 吗？比如 list_append([[1,2,3]], [2])

File: tools/python_bind/tests/test_db_list.py [hunk 2]
```diff
@@ -29,6 +29,109 @@ def _nested_list(value):
...
+            "CAST([2, CAST(NULL, 'INT64')], 'INT64[]'));",
+            [1, None, 2, None],
+        ),
+    ]
+    for query, expected in cases:
+        value = list(conn.execute(query))[0][0]
+        assert _nested_list(value) == expected
+
+    assert list(conn.execute("RETURN list_append(CAST(NULL, 'INT64[]'), 3);")) == [
+        [None]
```
Comment (longbinlai) [inline tools/python_bind/tests/test_db_list.py]: 这里很怪，到底 NULL, NONE 是什么意思，为什么输入是 NULL, 输出却变成 NONE

Comment (shirly121) [inline tools/python_bind/tests/test_db_list.py:81]: 实现层面就是用 None 来代表 NULL 值，NULL 是语法层面的表示，这个应该是 kuzu 语法和实现层面没统一导致的问题

File: doc/source/cypher_manual/expression/list_func.md (hunk 1 above)
Comment (longbinlai) [reply doc/source/cypher_manual/expression/list_func.md:9]: 补充下我自己的分析（未实测，请作者确认）：

类型推导机制上应该支持嵌套 —— tryGetMaxLogicalType 对嵌套类型走 tryCombineListTypes 递归合并子类型（src/compiler/common/types/types.cpp），`[[1,2,3]]` 是 LIST<LIST<INT64>>，append 的 `[2]` 是 LIST<INT64>，应能推出 LIST<LIST<INT64>>。嵌套列表值和嵌套字面量（`[[1,2],[3,4]]`）在现有测试里也已有覆盖。

所以预期：
- `list_append([[1,2,3]], [2])` → `[[1,2,3],[2]]`
- `list_append([1,2], [3])`（往 INT 列表里 append 一个列表）→ 报 "cannot find a common element type"

麻烦补一个嵌套用例的测试确认，并在文档加一句嵌套列表的说明。

File: tools/python_bind/tests/test_db_list.py (hunk 2 above)
Comment (longbinlai) [reply tools/python_bind/tests/test_db_list.py:81]: 自答下这行的疑问，行为本身是对的，"怪"是三层原因叠加：

1. NULL（Cypher）和 None（Python）是同一个东西：Python 绑定把数据库 NULL 转成 None 返回。`[[None]]` = 1 行 × 1 列，该列值是 NULL。
2. 语义是 null 传播：输入 NULL 列表 → 输出 NULL 列表（不是 [3]，也不是报错）。DuckDB 的 list_append(NULL, 3) 同样返回 NULL，Neo4j 的 null + list 也是 null。
3. 之所以要 CAST(NULL, 'INT64[]')：裸 NULL 的类型是 UNKNOWN，而 bindAppend 要求第一参数必须是 LIST/ARRAY，裸 NULL 会直接 binder 报错，测试只能先 CAST 出具体类型。

两个建议：
- 测试里加一行注释说明这组断言验证的是 null 传播语义；文档 list_func.md 也补一节 "NULL 列表输入返回 NULL"（目前文档只写了 null 元素追加和落库行为，没写 null 输入）。
- 更实质的一点：`list_append(NULL, 3)` 裸 NULL 直接报错对用户不友好，可以考虑 bind 期对 kUnknown（裸 NULL）放行、直接返回 NULL（DuckDB 行为）。

File: include/neug/compiler/function/list/functions/list_append_function.h [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+#pragma once
+
+#include "neug/common/types/value.h"
+
+namespace neug {
+namespace function {
+
+struct ListAppend {
+  static neug::Value operation(const std::vector<neug::Value>& args);
```
[the remark this answers is not in the document]
Comment (shirly121) [reply include/neug/compiler/function/list/functions/list_append_function.h:25]: Skipped this suggestion after compile verification. In this codebase `Value` is declared directly in namespace `neug` (`include/neug/common/types/value.h`), not in `neug::common`; changing the signature to `neug::common::Value` fails to compile. The existing explicit `neug::Value` signature is therefore correct.

File: include/neug/common/columns/list_columns.h [hunk 4]
```diff
@@ -133,7 +133,14 @@ class ListColumn : public IContextColumn {
...
   explicit ListColumnBuilder(DataType type) : type_(type), cur_offset_(0) {
-    child_builder_ = ColumnsUtils::create_builder(type_);
+    // LIST<UNKNOWN> is valid for an empty list literal. It has no materialized
+    // child values, but still needs a concrete empty data column so the result
+    // column can be finalized and returned to clients.
+    if (type_.id() == DataTypeId::kUnknown) {
+      child_builder_ = std::make_shared<ValueColumnBuilder<std::string>>();
+    } else {
+      child_builder_ = ColumnsUtils::create_builder(type_);
+    }
```
[the remark this answers is not in the document]
Comment (shirly121) [reply include/neug/common/columns/list_columns.h:143]: I investigated this with a dedicated UNKNOWN column/builder and added a regression case for `list_append([], NULL)`. The dedicated UNKNOWN column causes the current result serialization path to abort because downstream materialization requires a concrete child column. The STRING builder is only an internal carrier for empty/untyped-NULL children; the parent `ListColumn` retains the logical `LIST<UNKNOWN>` type, and the regression test passes. I therefore kept the existing implementation rather than introducing a breaking UNKNOWN column.

File: src/compiler/function/list/list_append_function.cpp [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+                           ArrayType::GetNumElements(inputType));
+  }
+  return DataType::List(elementType.copy());
+}
+
+const std::vector<Value>& getChildren(const Value& value) {
+  return value.type().id() == DataTypeId::kList
+             ? ListValue::GetChildren(value)
+             : ArrayValue::GetChildren(value);
+}
```
[the remark this answers is not in the document]
Comment (shirly121) [reply src/compiler/function/list/list_append_function.cpp]: Fixed in ef2cb9d7. The shared list/array helpers are now in `list_function_utils.h` and are used by both list_append and list_concat.

File: doc/source/cypher_manual/expression/list_func.md (hunk 1 above)
Comment (shirly121) [reply doc/source/cypher_manual/expression/list_func.md:9]: 已确认支持 list of list，并在 ef2cb9d7 中补充了 Python 端到端测试和文档示例：`list_append([[1, 2], [3, 4]], [5, 6])` 返回 `[[1, 2], [3, 4], [5, 6]]`。文档在 list_append/list_concat 开头也明确说明支持类型兼容的嵌套 LIST/ARRAY。

Comment (shirly121) [reply doc/source/cypher_manual/expression/list_func.md:9]: 已按建议实测并补充两类测试：兼容的嵌套追加/拼接均通过；`list_append([1, 2], [3])` 会在 binder 阶段报 `cannot find a common element type`，对应的异常断言也已加入 test_db_list.py。

File: tools/python_bind/tests/test_db_list.py (hunk 2 above)
Comment (shirly121) [reply tools/python_bind/tests/test_db_list.py:81]: 已补充 typed NULL list 的传播注释和文档示例，明确返回的是 NULL；同时保留现有裸 NULL 绑定语义。放宽 `list_append(NULL, 3)` 涉及函数签名推导和通用 UNKNOWN 语义，不属于本次 list append/concat 功能范围，因此本 PR 不扩展该行为。

File: src/compiler/gopt/g_expr_converter.cpp [hunk 6]
```diff
@@ -694,10 +694,6 @@ std::unique_ptr<::common::Expression> GExprConverter::convertToTupleFunc(
 std::unique_ptr<::common::Expression> GExprConverter::convertToListFunc(
     const binder::Expression& expr,
     const std::vector<std::string>& schemaAlias) {
-  if (expr.getChildren().empty()) {
-    THROW_EXCEPTION_WITH_FILE_LINE(
-        "Array function should have at least one child");
-  }
   auto listPB = std::make_unique<::common::ToList>();
```
Comment (longbinlai) [inline src/compiler/gopt/g_expr_converter.cpp:697]: 这里删掉 "Array function should have at least one child" 的异常是本 PR 一个关键改动，但它实际服务的场景目前没有任何测试覆盖，建议补用例。

机制上分两条路径：

1. **现有测试全部走常量折叠路径**：`RETURN list_append([], 1)`、`list_concat([], [])` 这类用例参数全是常量，bind 期就被折叠（`ConstantExpressionVisitor::needFold` → `foldExpression`）成 literal，根本不会走到 convertToListFunc。

2. **真正依赖这个改动的是不折叠场景**：比如 `MATCH (item:Item) RETURN list_append([], item.id)` —— 属性参数让整个函数表达式不可折叠，空列表参数在 expression_binder.cpp 里经 forceCast 变成 `CAST(LIST_CREATION(), LIST<INT64>)`，转换到这里就是 0 个子节点的 LIST_CREATION：改动前会直接抛 "Array function should have at least one child"，现在生成一个空 ToList 进入物理执行层。

也就是说，**空 ToList 在运行时能否正确产生空列表值，是本 PR 引入的全新执行路径，目前没有被任何测试执行过**（我只做了代码路径分析，未实测运行时行为）。

建议在 test_list_func_docs.py 补一组不折叠用例，例如：

```python
# item.id 是属性参数，整条表达式不可常量折叠，空列表会以空 ToList 进入运行时
assert _scalar(conn, "MATCH (item:Item) RETURN list_append([], item.id);") == [1]
assert _scalar(conn, "MATCH (item:Item) RETURN list_concat([], item.tags);") == [1, 2]
```

如果空 ToList 在执行层有坑（返回 null / 报错 / 行为异常），这组用例正好能暴露出来。

Comment (shirly121) [reply src/compiler/gopt/g_expr_converter.cpp:697]: 已在 bf4f8dff 中补充两条非折叠端到端用例：`list_append([], item.id)` 和 `list_concat([], item.tags)`。属性参数会阻止常量折叠，确保空列表以零子节点 ToList 进入物理执行层；两条用例均已实测通过，目标文档测试共 3 passed。

**Review by longbinlai — APPROVED** [review]
LGTM

File: include/neug/common/columns/list_columns.h [hunk 7]
```diff
@@ -133,7 +133,14 @@ class ListColumn : public IContextColumn {
 class ListColumnBuilder : public IContextColumnBuilder {
  public:
   explicit ListColumnBuilder(DataType type) : type_(type), cur_offset_(0) {
-    child_builder_ = ColumnsUtils::create_builder(type_);
+    // LIST<UNKNOWN> is valid for an empty list literal. It has no materialized
+    // child values, but still needs a concrete empty data column so the result
+    // column can be finalized and returned to clients.
+    if (type_.id() == DataTypeId::kUnknown) {
```
Comment (liulx20) [inline include/neug/common/columns/list_columns.h:139]: [P1] LIST<UNKNOWN> 可以返回，但无法被 UNWIND 消费
[list_columns.h#L135-L143](https://github.com/alibaba/neug/blob/b1875f2e653c73118ebfe57e1721ff2641e49f3b/include/neug/common/columns/list_columns.h#L135-L143) 将 UNKNOWN 临时映射成字符串 column builder，只解决了最终结果序列化，没有让下游算子支持这种类型。
本地复现：
UNWIND list_append([], NULL) AS value RETURN value;
报错：
Unsupported data type for column builder: 2
物化后仍然失败：
WITH list_append([], NULL) AS values
UNWIND values AS value
RETURN value;
not implemented for ListColumn[1] 2
list_append([], NULL) 是 PR 明确支持且已有测试的输入，因此其结果也应该能参与后续查询。建议统一处理 UNKNOWN/NULL 子列，并补充上述两类 UNWIND 测试。

这个有更好的处理方式吗？
